lighthouse-viewer/app/src/main.js

```diff
@@ -0,0 +1,505 @@
+    console.error(msg);
+  }
+
+  /**
+   * Explicitly hides the butter bar.
+   */
+  hide() {
+    clearTimeout(this._id);
+    this.el.classList.remove('show');
+  }
+}
+
+const logger = new Logger('#log');
+
+class FirebaseAuth {
+  constructor() {
+    this.accessToken = null;
+    this.user = null;
+
+    this.provider = new firebase.auth.GithubAuthProvider();
+    this.provider.addScope('gist');
+
+    firebase.initializeApp({
+      apiKey: 'AIzaSyApMz8FHTyJNqqUtA51tik5Mro8j-2qMcM',
+      authDomain: 'lighthouse-viewer.firebaseapp.com',
+      databaseURL: 'https://lighthouse-viewer.firebaseio.com',
+      storageBucket: 'lighthouse-viewer.appspot.com',
+      messagingSenderId: '962507201498'
+    });
+
+    // Wrap auth state callback in a promise so  other parts of the app that
+    // require login can hook into the changes.
+    this.ready = new Promise((resolve, reject) => {
+      firebase.auth().onAuthStateChanged(user => {
+        idb.get('accessToken').then(token => {
+          if (user && token) {
+            this.accessToken = token;
+            this.user = user;
+          }
+          resolve(user);
+        });
+      });
+    });
+  }
+
+  /**
+   * Signs in the user to Github using the Firebase API.
+   * @return {!Promise<object>} The logged in user.
+   */
+  signIn() {
+    return firebase.auth().signInWithPopup(this.provider).then(result => {
+      this.accessToken = result.credential.accessToken;
+      // A limitation of FB auth is that it doesn't return an oauth token
+      // after a page refresh. We'll get a firebase token, but not an oauth token
+      // for GH. Since GH's tokens never expire, stash the access token in IDB.
+      idb.set('accessToken', this.accessToken); // Note: async.
+      this.user = result.user;
+      return this.accessToken;
+    });
+  }
+
+  /**
+   * Signs the user out.
+   * @return {!Promise}
+   */
+  signOut() {
+    return firebase.auth().signOut().then(() => {
+      this.accessToken = null;
+      idb.delete('accessToken'); // Note: async.
+    });
+  }
+}
+
+class GithubAPI {
+  constructor() {
+    // this.CLIENT_ID = '48e4c3145c4978268ecb';
+    this.auth = new FirebaseAuth();
+  }
+
+  static get LH_JSON_EXT() {
+    return '.lighthouse.report.json';
+  }
+
+  authorize() {
+    return this.auth.accessToken ? Promise.resolve(this.auth.accessToken) :
+                                   this.auth.signIn();
+  }
+
+  /**
+   * Creates a gist under the users account.
+   * @param {!string} jsonFile The gist file body.
+   * @return {!Promise<string>} id of the created gist.
+   */
+  createGist(jsonFile) {
+    logger.log('Saving JSON to Github gist...', false);
+
+    return this.authorize()
+      .then(accessToken => {
+        // Stringify twice so quotes are escaped for POST request to succeed.
+        const filename = AssetSaver.getFilenamePrefix({url: jsonFile.url});
+        const content = JSON.stringify(JSON.stringify(jsonFile));
+        const body = `{
```

Comment:
why the string body and early double `stringify` instead of just object stringified at the end?